### PR TITLE
Use `similar` in `empty`

### PR DESCRIPTION
**Pull request by jishnub**
After this,
```julia
julia> S = StructArray{Complex{Int}}(([1,2], [3,4]))
2-element StructArray(::Vector{Int64}, ::Vector{Int64}) with eltype Complex{Int64}:
 1 + 3im
 2 + 4im

julia> empty(S)
0-element StructArray(::Vector{Int64}, ::Vector{Int64}) with eltype Complex{Int64}
```
instead of
```julia
julia> empty(S)
Complex{Int64}[]
```
This behavior matches the docstring now: "Create an empty vector similar to `v`". Often, `similar` will fall back to creating a `Vector`, so the current behavior will be preserved.

~I'm unsure about what test to add for this, so suggestions are welcome.~ Test added for a `StructArray` through a new test helper

### CODE REVIEW

**Comment by Seelengrab** [comment]
Perhaps a test with the OffsetArray testhelper?

File: base/abstractarray.jl [hunk 1]
```diff
@@ -888,10 +888,11 @@ julia> empty([1.0, 2.0, 3.0], String)
 String[]
 ```
 """
-empty(a::AbstractVector{T}, ::Type{U}=T) where {T,U} = Vector{U}()
+empty(a::AbstractVector{T}, ::Type{U}=T) where {T,U} = similar(a, U, 0)
 
-# like empty, but should return a mutable collection, a Vector by default
-emptymutable(a::AbstractVector{T}, ::Type{U}=T) where {T,U} = Vector{U}()
+# like empty, but should return a mutable collection
+emptymutable(a::AbstractVector{T}, ::Type{U}=T) where {T,U} = similar(a, U, 0)
```
Comment (N5N3) [inline base/abstractarray.jl:894]: L891 looks OK to me.
But I'm not sure if L894 is breaking. (`emptymutable` is used in `union` and the general fallback is `push!` based.)

Comment (jishnub) [reply base/abstractarray.jl]: I've reverted the `emptymutable` change, so only L891 remains now

**Comment by jishnub** [comment]
Gentle bump

**Comment by N5N3** [comment]
[This list](https://juliahub.com/ui/Search?q=empty&type=symbols&u=use) is too long to go through.
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/49958#issuecomment-1889252078) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/999a140_vs_f345755/report.html) is available.

**Pull request merged**